### PR TITLE
Render and manage reservations that reserve no address

### PR DESCRIPTION
**Pull request by marcinpsk**
Fixes #110

A Kea host reservation may legally reserve no address — an identifier-only DHCPv4 host carrying just a hostname, option-data or client-classes, or a DHCPv6 host that only delegates prefixes. The reservations tab crashed rendering such rows (every mutation route was keyed by IP) and the sync job failed on them.

Commit 1 stops the bleeding: the tab renders again and the sync job no longer fails on an address-less host.

Commit 2 makes address-less reservations first-class:

- **Tables**: identifier column showing the value and which Kea key it came from (previously only hw-address/DUID rendered; other identifier types showed a blank cell), a v6 delegated-prefixes column, and a muted "No address" badge instead of an empty cell. Global tabs inherit all of it.
- **Edit/delete by client identifier**: four integer-only routes carrying the identifier in the query string, so no reservation value can make a URL fail to reverse; one shared lookup validates GET and POST identically (unknown/wrong-version type, missing, empty, repeated, conflicting or over-long values → 400; well-formed but absent in Kea → 404). The address stays editable, so a host can gain or lose its fixed address without changing the key it is addressed by.
- **Forms/CSV**: address optional everywhere; per-type identifier validation shared between form and CSV importer, with type-aware length caps (a maximal 128-octet DUID is 383 characters and is now accepted; opaque identifiers keep the 255 guard); a shared delegated-prefix validator (canonical form, deduplicated, bounded, prefix length ≥ 1); CSV columns matched by header name with exactly one identifier column per file.
- **Lease status without an address**: lease enrichment also collects per-subnet lease identifiers (v4 hw-address/client-id, v6 duid/hw-address), so an address-less row shows its client's active lease — matched only within its own subnet. Addressed rows keep the IP match.
- An options formset absent from a submission no longer wipes option-data the request never touched; a submitted formset with all rows deleted still clears them.
- `reservation_deleted` now sends a uniform payload on every route and outcome (`ip_address` / `identifier_type` / `identifier`, `None` where not applicable).

## Testing

- Unit: **2403 passed** (27 new tests for the behaviors above, each proven red against the unfixed code).
- Integration against real Kea 3.2.0 daemons: **149 passed**, including 5 new tests that drive the add / edit-by-identifier / delete-by-identifier pages and read the host back from the daemon — proving real Kea accepts the payloads this PR builds, and that untouched option-data survives an identifier-keyed `reservation-update`. The test harness gained a PostgreSQL hosts database (`libdhcp_pgsql` + `libdhcp_host_cmds`, schema loaded by a one-shot init service from the pinned upstream `dhcpdb_create.pgsql`) because `host_cmds` persists only to a host database.
- ruff, mock-discipline, REUSE and all pre-commit hooks clean.

## Summary by CodeRabbit

* **New Features**
  * Added identifier-keyed edit/delete routes and workflows for address-less DHCPv4 reservations and DHCPv6 prefix-only reservations.
  * Updated reservation UI tables to display identifier/prefix details with reliable “No address” rendering.
* **Bug Fixes**
  * Improved IPAM sync summaries and conflict reporting, including accurate skipped counts for no-address reservations.
* **Documentation**
  * Expanded README for sync summary fields and reservation delete signal payload behavior.
* **Tests**
  * Added/expanded coverage for bulk import, UI flows, validation edge cases, and conflict/skip scenarios.

### CODE REVIEW

File: tests/test_setup.sh [hunk 1]
```diff
@@ -8,6 +8,19 @@ openssl req -new -newkey rsa:2048 -sha256 -days 365 -nodes -x509 -keyout ./tests
...
+# release tarball of the version the daemons run (see docker-compose.override.yml).
+KEA_VERSION="${KEA_VERSION:-3.2.0}"
+KEA_TARBALL_SHA256="${KEA_TARBALL_SHA256:-14bf695d37b65b9b1bf550fea5d0adaf9806c50e5419ef2a176a4b8e9aade3df}"
+echo "Fetching Kea $KEA_VERSION host database schema"
+mkdir -p ./tests/docker/kea_schema/
+curl -fsSL -o "/tmp/kea-$KEA_VERSION.tar.xz" \
+    "https://downloads.isc.org/isc/kea/$KEA_VERSION/kea-$KEA_VERSION.tar.xz"
+echo "$KEA_TARBALL_SHA256  /tmp/kea-$KEA_VERSION.tar.xz" | sha256sum -c -
+tar xJf "/tmp/kea-$KEA_VERSION.tar.xz" -C ./tests/docker/kea_schema/ --strip-components=6 \
+    "kea-$KEA_VERSION/src/share/database/scripts/pgsql/dhcpdb_create.pgsql"
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply tests/test_setup.sh]: > **Use an unpredictable temporary path.** `/tmp/kea-$KEA_VERSION.tar.xz` is predictable and can be pre-created or symlinked by another local user before `curl` writes it.

Fixed in b40b069: the tarball now downloads into a `mktemp -d` directory removed by an EXIT trap; the pinned SHA256 verification and exact-member extraction are unchanged. Validated by running the fetch/verify/extract block standalone (checksum matched, `dhcpdb_create.pgsql` extracted, temp dir cleaned up).

File: netbox_kea/jobs.py [hunk 2]
```diff
@@ -183,7 +218,13 @@ def _sync_server_leases(
...
             had_errors = True
 
-    stats["conflicts"] = stats.get("conflicts", 0) + len(conflicts)
+    # Deduplicated across phases and versions by the caller's set; the accumulator
+    # itself stays a list because sync_lease_to_netbox appends to it.
+    if conflict_ips is not None:
+        conflict_ips.update(_canonical_ip(ip) for ip in conflicts)
+        stats["conflicts"] = len(conflict_ips)
+    else:
+        stats["conflicts"] = stats.get("conflicts", 0) + len(conflicts)
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_kea/jobs.py]: > **Duplicated conflict-dedup/stat-update snippet.** The block appears verbatim in both `_sync_server_leases` and `_sync_server_reservations`'s `finally`.

Fixed in 2a4e93d: extracted `_record_conflicts(stats, conflicts, conflict_ips)` and replaced both call sites (the reservation one stays inside its `finally`). Added `TestRecordConflicts` covering the dedup-set and accumulate branches; the existing jobs tests stay green.

File: netbox_kea/tables.py [hunk 3]
```diff
@@ -515,24 +515,22 @@ class Meta(NetBoxTable.Meta):
...
-{% if record.can_change %}
-<span class="btn-group">
-  <a href="{% url "plugins:netbox_kea:server_reservation6_edit" record.server_pk record.subnet_id record.ip_address %}"
-     class="btn btn-sm btn-warning" aria-label="Edit reservation {{ record.ip_address }}"><i class="mdi mdi-pencil" aria-hidden="true"></i></a>
-  <a href="{% url "plugins:netbox_kea:server_reservation6_delete" record.server_pk record.subnet_id record.ip_address %}"
+  {% endif %}
+  {% if record.delete_url %}
+  <a href="{{ record.delete_url }}"
      class="btn btn-sm btn-danger" aria-label="Delete reservation {{ record.ip_address }}"><i class="mdi mdi-trash-can-outline" aria-hidden="true"></i></a>
+  {% endif %}
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_kea/tables.py:533]: > **Action buttons lose their accessible name on exactly the rows this PR adds.** `aria-label="Edit reservation {{ record.ip_address }}"` renders as `Edit reservation ` for an address-less reservation.

Fixed in 4dfba4f: both edit and delete labels now use `{{ record.ip_address|default:record.identifier }}`. Added `test_address_less_row_labels_its_actions_with_the_identifier` (red before the fix: the rendered label was empty) plus a guard that an addressed row still labels with its IP.

File: netbox_kea/tests/test_utilities.py [hunk 4]
```diff
@@ -237,6 +238,52 @@ def test_mixed_case_accepted(self):
...
+    identifier (circuit-id / flex-id / remote-id) out of the Kea API.
+    """
+
+    def _hex(self, octets: int) -> str:
+        return ":".join(["ab"] * octets)
+
+    def _validate(self, identifier_type: str, value: str) -> str:
+        from netbox_kea.utilities import validate_reservation_identifier
+
+        return validate_reservation_identifier(identifier_type, value)
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_kea/tests/test_utilities.py]: > **Use `SimpleTestCase` for this DB-independent helper test class.**

Fixed in fc16ec8: `TestValidateReservationIdentifierLength` now inherits `SimpleTestCase`, so a future DB query in the helper fails loudly instead of silently turning it into a DB test.

File: netbox_kea/tests/test_views_combined.py [hunk 5]
```diff
@@ -116,3 +116,96 @@ def test_multi_page_pagination_followed(self):
...
+        return reverse(f"plugins:netbox_kea:combined_reservations{version}") + f"?server={self.server.pk}"
+
+    def test_v4_identifier_only_reservation_renders(self):
+        page = {
+            "result": 0,
+            "arguments": {
+                "hosts": [{"subnet-id": 3742, "hw-address": "aa:bb:cc:dd:ee:ff", "hostname": "printer-1"}],
+                "next": {"from": 0, "source-index": 0},
+            },
+        }
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_kea/tests/test_views_combined.py]: > **Reuse `_res_page` from `kea_stub` instead of re-shaping the page payload here.**

Fixed in fc16ec8: the local `_page` staticmethod and both inline page dicts are replaced by `_res_page` imported from `.kea_stub`; the affected combined-view tests stay green.

File: netbox_kea/utilities.py [hunk 6]
```diff
@@ -314,40 +481,26 @@ def parse_reservation_csv(content: str, version: int) -> list[dict[str, Any]]:
 
     rows: list[dict[str, Any]] = []
     for row_num, raw in enumerate(reader, start=2):  # header is row 1
+        # DictReader files cells with no header under the None key.  They carry data
+        # this parser never reads, so reject the row rather than drop them.
+        if raw.get(None):
+            raise ValueError(f"Row {row_num}: more values than the header has columns")
         row = {k.strip(): (v or "").strip() for k, v in raw.items() if k is not None}
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_kea/utilities.py]: > **A trailing comma now fails the whole import.** `csv.DictReader` collects surplus cells into a list under the `None` key, so a row ending in a stray comma yields `[""]` — truthy — and the row is rejected.

Fixed in 4dfba4f: the check now rejects only when a surplus cell carries a non-empty value (`any((v or "").strip() for v in raw.get(None) or [])`). Red→green tests: a row with a trailing comma imports; a row with a genuinely filled surplus cell is still rejected.

File: README.md [hunk 7]
```diff
@@ -242,6 +251,18 @@ The `Kea IPAM Sync` job runs automatically when `rqworker` is active:
 5. One server failing does not block others
 6. Summary logged per server and in total
 
+Each server's summary reports `synced`, `skipped`, `errors` and `conflicts`:
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply README.md]: > **`synced` isn't an actual summary field.**

Fixed in cc3db16: the README sentence now names the keys the job actually reports (`created`, `updated`, `errors`, `prefix_errors`, `conflicts`, `skipped`), verified against `jobs.py`'s `run()` summary.

File: tests/conftest.py [hunk 8]
```diff
@@ -1,11 +1,14 @@
...
 import pynetbox
 import pytest
 import requests
 
 
 @pytest.fixture(scope="session")
 def netbox_url() -> str:
-    return "http://localhost:8000"
+    # Overridable so the harness can be published on another port when 8000 is taken.
+    return os.environ.get("NETBOX_URL", "http://localhost:8000")
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply tests/conftest.py]: > **Add the required SPDX header to this Python source file.** `tests/conftest.py` starts with `import os` and contains no inline SPDX license header.

Declined: this file is already licensed through `REUSE.toml`, whose `tests/**` annotation (lines 14–19, Apache-2.0) covers it — the same mechanism AGENTS.md endorses alongside inline headers, and the reason `reuse lint` passes (184/184 files compliant; it runs in pre-commit on every commit here). The file also predates this PR without an inline header on `main`; this PR only added a fixture to it. Converting the `tests/` tree to inline headers would be an unrelated repo-wide change.

[the remark this answers is not in the document]
Comment (marcinpsk) [reply tests/conftest.py]: > **Fall back to the default `NETBOX_URL` when the env var is empty.**

Fixed in b40b069: `netbox_url()` now treats an empty variable as unset (`os.environ.get("NETBOX_URL") or default`) and strips trailing slashes so appended paths don't double the `/`. Validated across unset, empty, and trailing-slash values.

File: tests/docker/docker-compose.override.yml [hunk 9]
```diff
@@ -6,9 +6,36 @@
...
+  # Both daemons share one, so the schema is loaded once before either starts.
+  # test_setup.sh fetches the schema for the exact Kea version the daemons run.
+  kea-db-init:
+    image: docker.io/postgres:16-alpine
+    depends_on:
+      - postgres
+    environment:
+      PGHOST: postgres
+      PGUSER: netbox
+      PGPASSWORD: J5brHrAXFLQSif0K
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply tests/docker/docker-compose.override.yml:20]: > **Externalize the shared test database credential.** The same database password is committed in the initializer and both Kea daemon configurations.

Declined: this is a throwaway credential for the ephemeral CI test database, in a compose file that already carries committed test credentials from upstream on `main` (`DB_PASSWORD: J5brHrAXFLQSif0K`, both Redis passwords, the `admin` superuser — `tests/docker/docker-compose.yml` lines 23–44). The postgres service publishes no host port (only NetBox's 8000 and the loopback-bound Kea sockets are exposed), the database holds generated test data only, and the stack lives for one CI run. Externalizing one credential beside its committed siblings would add templating machinery to the harness without changing its security posture.

File: netbox_kea/views/reservations.py [hunk 10]
```diff
@@ -1024,13 +1285,13 @@ def post(self, request: HttpRequest, pk: int, subnet_id: int, ip_address: str) -
                 )
                 return redirect(return_url)
             except KeaException as exc:
-                logger.exception("Failed to update DHCPv6 reservation for %s", ip_address)
+                logger.exception("Failed to update DHCPv6 reservation for %s", lookup.label)
```
Comment (marcinpsk) [inline netbox_kea/views/reservations.py:1288]: @coderabbitai Outside-diff finding:

> **DHCPv6 edit still logs `ip_address`, which is empty on the identifier-keyed route.** (`netbox_kea/views/reservations.py` GET ~1159/1163, POST error paths ~1288-1296)

Fixed in 4dfba4f: all five v6 log calls now use `lookup.label` (the v4 views already did), so an identifier-keyed failure logs `duid <value>` instead of an empty subject; the now-unused local `ip_address` in `post()` is removed. Red→green tests `test_v6_update_failure_logs_the_identifier` and `test_v6_fetch_failure_logs_the_identifier` assert the identifier appears in the captured log records.

File: tests/conftest.py [hunk 11]
```diff
@@ -1,11 +1,17 @@
...
 
 
 @pytest.fixture(scope="session")
 def netbox_url() -> str:
-    return "http://localhost:8000"
+    # Overridable so the harness can be published on another port when 8000 is taken.
+    # An empty NETBOX_URL means "unset", and a trailing slash would double up the / in
+    # every f-string that appends a path.
+    url = os.environ.get("NETBOX_URL") or "http://localhost:8000"
+    return url.rstrip("/")
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply tests/conftest.py]: > **Treat whitespace-only `NETBOX_URL` values as unset.** `or` handles `""`, but not `"   "`.

Fixed in c6f3f3e: `os.environ.get("NETBOX_URL", "").strip() or "http://localhost:8000"` — validated across unset, empty, whitespace-only, and padded/trailing-slash values (whitespace-only now falls back; ` http://host:8010/ ` resolves to `http://host:8010`).

File: netbox_kea/tests/test_views_sync.py [hunk 12]
```diff
@@ -904,3 +904,118 @@ def test_mixed_success_skip_and_error_in_batch(self):
...
+
+    def test_one_bad_row_rejects_the_whole_file(self):
+        """The parser runs before any write, so a later bad row cannot half-import a file."""
+        upload = self._upload(
+            b"subnet-id,hw-address\n1,aa:bb:cc:dd:ee:ff\n1,not-a-mac\n",
+        )
+        with stub_kea({}) as kea:
+            response = self.client.post(self._url(), {"csv_file": upload})
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(kea.commands(), [])
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_kea/tests/test_views_sync.py:986]: > **Assert the parse failure, not only the lack of writes.** A broken implementation that silently drops malformed rows would still send no Kea commands and pass this test.

Fixed in 149be4c: the test now also asserts `assertContains(response, "parsing failed")`, matching the sibling invalid-input cases. Verified the assert bites (a deliberately wrong string fails on the real `Row 3: invalid hw-address` path).

File: netbox_kea/tests/test_views_sync.py [hunk 13]
```diff
@@ -904,3 +904,118 @@ def test_mixed_success_skip_and_error_in_batch(self):
...
+        from django.contrib.contenttypes.models import ContentType
+        from users.models import ObjectPermission
+
+        from netbox_kea.models import Server
+
+        restricted = User.objects.create_user(
+            username="noperms_import",
+            email="noperms_import@example.com",
+            password="pass",
+        )
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_kea/tests/test_views_sync.py:999]: > **Create test users through `get_user_model()`.** Use `get_user_model().objects.create_user(...)` at both sites rather than the concrete `User` model.

Declined: these sites already go through `get_user_model()`. The file never imports Django's concrete model — `test_views_sync.py:41` is `from .utils import _PLUGINS_CONFIG, User, _ViewTestBase`, and `tests/utils.py:18` defines `User = get_user_model()` (the shared alias the whole suite uses). Inlining `get_user_model()` at just these two call sites would diverge from the file's other usage and the `.utils` convention without changing behavior.

File: tests/conftest.py [hunk 14]
```diff
@@ -1,11 +1,17 @@
+import os
+
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply tests/conftest.py:2]: > **Add the required inline SPDX header.** This repeats the previous review finding.

Declined, same evidence as the prior thread on this file: `tests/conftest.py` is licensed via `REUSE.toml`'s `tests/**` annotation (Apache-2.0), the mechanism AGENTS.md endorses alongside inline headers, and `reuse lint` passes 184/184 in pre-commit on every commit. The file predates this PR without an inline header; converting the `tests/` tree to inline headers would be an unrelated repo-wide change.

File: tests/docker/docker-compose.override.yml [hunk 15]
```diff
@@ -6,9 +6,36 @@
...
+    entrypoint:
+      - sh
+      - -c
+      - |
+        set -e
+        until pg_isready -q; do sleep 1; done
+        psql -d postgres -tAc "SELECT 1 FROM pg_database WHERE datname='kea'" | grep -q 1 \
+          || psql -d postgres -c "CREATE DATABASE kea"
+        psql -d kea -tAc "SELECT 1 FROM pg_tables WHERE tablename='schema_version'" | grep -q 1 \
+          || psql -q -v ON_ERROR_STOP=1 -d kea -f /schema/dhcpdb_create.pgsql
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply tests/docker/docker-compose.override.yml:32]: > **Make the schema initialization check use `schema_version.version`/`minor`.** Checking only whether the table exists can skip `dhcpdb_create.pgsql` for an older, incompatible, or manually created `kea` database.

Declined: the scenario needs a reused postgres volume, and this stack doesn't have one in CI — the compose project starts from a fresh volume every run, so the database is always created by this init service. For a stale local volume, Kea 3.2.0 itself already fails loudly: the daemons validate `schema_version` at startup and refuse a mismatch (observed while building this harness — a 29.0 schema from kea-admin 3.0.3 was rejected by the 3.2.0 daemons, which is why the schema is fetched from the matching upstream tarball at all). A second version check in the init service would only move the same loud failure one service earlier.

File: tests/test_setup.sh [hunk 16]
```diff
@@ -8,6 +8,24 @@ openssl req -new -newkey rsa:2048 -sha256 -days 365 -nodes -x509 -keyout ./tests
 openssl req -new -newkey rsa:2048 -sha256 -days 365 -nodes -x509 -keyout ./tests/docker/certs/nginx.key -out ./tests/docker/certs/nginx.crt -addext "subjectAltName=DNS:nginx" -subj "/CN=nginx"
 chmod -R 0777 ./tests/docker/certs/
 
+# Reservation management needs a Kea host database, and Kea refuses a schema whose
+# version is not exactly the one it was built against, so take the schema from the
+# release tarball of the version the daemons run (see docker-compose.override.yml).
+KEA_VERSION="${KEA_VERSION:-3.2.0}"
+KEA_TARBALL_SHA256="${KEA_TARBALL_SHA256:-14bf695d37b65b9b1bf550fea5d0adaf9806c50e5419ef2a176a4b8e9aade3df}"
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply tests/test_setup.sh]: > **Tie the fetched schema, checksum, and daemon images to one Kea version.**

Fixed in 06ce8d0: the compose image tags now interpolate `${KEA_VERSION:-3.2.0}` and `test_setup.sh` exports `KEA_VERSION`, so the tarball, schema and both daemon images derive from the single version source. Verified via `docker compose config`: unset → both tags 3.2.0; `KEA_VERSION=9.9.9` → both tags 9.9.9. `KEA_TARBALL_SHA256` stays fail-closed on purpose — overriding the version without the matching checksum fails `sha256sum -c` loudly rather than running with unverified artifacts.

File: tests/test_ui.py [hunk 17]
```diff
@@ -1631,3 +1631,273 @@ def search() -> None:
...
+    kea_client.command(
+        "lease4-add",
+        service=["dhcp4"],
+        arguments={"ip-address": "192.0.2.150", "hw-address": mac, "hostname": "leased-client"},
+    )
+    _add_reservation4_without_address(page, plugin_base, reservation_server.id, mac, "no-address-leased")
+
+    row = page.locator("table.object-list > tbody > tr").filter(has_text=mac)
+    expect(row).to_have_count(1)
+    expect(row.get_by_text("Active Lease", exact=True)).to_be_visible()
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply tests/test_ui.py]: > **Clean up the lease this test injects into the shared daemon.**

Fixed in d7f6dfe: the post-injection body is wrapped in `try`/`finally` and the `finally` issues `lease4-del` for the injected address. Validated against the live stack: with the pre-fix file, `lease4-get 192.0.2.150` after the test returned the leaked lease (`result 0`); with the fix, `result 3` "Lease not found" — and the address sits inside the `leases4_250` fixture range other tests use, so the leak was consequential.

File: tests/conftest.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (marcinpsk) [reply tests/conftest.py:2]: > To prevent future contradictory findings, consider explicitly documenting tests/** as an allowed REUSE.toml annotation exception in AGENTS.md or the review guidance.

Done in 603b5ed: the AGENTS.md REUSE/SPDX bullet now states the split explicitly — new source files carry inline headers, while the upstream-inherited trees (`netbox_kea/**`, `tests/**`) and docs/config are bulk-annotated in `REUSE.toml`, and a file covered there needs no inline header. Thanks for the suggestion.
